### PR TITLE
Newsfeed

### PR DESCRIPTION
**Pull request by ericljpemberton**
#### Functionality

Should look correct on pretty much any device.
Tested on Nexus 5x, Nexus 6p, iphone 5, iphone 6, iphone 6 plus, ipad, galaxy note 3 as well as a variety of laptop sizes

#### Todo's

I need to determine if a user has connected Facebook and/or Twitter (How can I find this info)
I need to pop up a window to connect to them if they click on them when trying to make a post (do we have a component for this)
After Kyle changes the API, I will be adding the user's post to the top of their feed so that can see that it was successfully posted

EDIT: DONE... I need to create a full screen photo view window to see news feed posts
Strictly limit characters to 140 for tweets. There is already a warning if a user is linking twitter and the post is over the limit but it needs to be enforced (cut off the text if it is over 140?)
EDIT: DONE... Currently unable to add offset and limits to newsfeed queries (talking to Kyle atm)
EDIT: DONE (A basic proof on concept has been created.... (Retweets need to be implemented 
EDIT DONE: Currently Photo is not added... Depending on how we use photo urls, this could also be included (Are photo URLs deleted if the post is deleted?)
EDIT DONE: I still need to divide things into more components...

In the future there should probably be a spinner for when we post as well as a spinner for pagination and loading a photo

### CODE REVIEW

**Comment by ericljpemberton** [comment]
Here are some screenshots of what it should hopefully look like...
Regular view
Very small laptop
Galaxy Note 3
iphone 6
Nexus 5x
iphone 5

**Comment by ericljpemberton** [comment]
Some may look a little weird but that's because some of my test text was literally one character and Next to the retweet button will be a show image button

File: src/js/organizer.client.jsx [hunk 1]
```diff
@@ -3,15 +3,15 @@ import ReactDOM from "react-dom";
...
+import Home from "pages/attendee/Home";
+import NewsFeed from "pages/attendee/NewsFeed";
+import People from "pages/attendee/People";
+import Events from "pages/attendee/Events";
+import EventPage from "pages/attendee/Event";
 import Profile from "pages/Profile";
-import Info from "pages/Info";
-import Login from "pages/Login";
+import Info from "pages/attendee/Info";
+import Login from "pages/attendee/Login";
```
Comment (ericljpemberton) [inline src/js/organizer.client.jsx:14]: Reminder to self to remove

Comment (larajanecka) [reply src/js/organizer.client.jsx:14]: The changes in the organizer PR should fix this.

File: src/js/pages/attendee/NewsFeed.jsx [hunk 2]
```diff
@@ -1,130 +1,210 @@
...
+// import style from '../../sass/newsfeed.scss';
 
+// TODO: LIMIT NUMBER OF NEWLINES?
 
 export default class NewsFeed extends React.Component {
 
-	constructor() {
+  constructor() {
     super();
     this.getNewsFeedPosts = this.getNewsFeedPosts.bind(this);
```
Comment (larajanecka) [inline src/js/pages/attendee/NewsFeed.jsx:12]: At some point we should switch to using fat arrow functions instead of all this binding.

File: src/js/pages/attendee/NewsFeed.jsx [hunk 3]
```diff
@@ -1,130 +1,210 @@
...
+        items: [],
+        nextOffset: null,
+      },
+      file: null,
+      preview: null,
       fbPost: false,
       twPost: false,
-      userText: "#SoftwareDemoDay",
+      // TODO: HOW CAN I GET THESE? (If user is integrated into social media channels)
+      twIntegration: true,
```
Comment (larajanecka) [inline src/js/pages/attendee/NewsFeed.jsx]: Might be a value inside auth0.

File: src/js/pages/attendee/NewsFeed.jsx [hunk 4]
```diff
@@ -1,130 +1,210 @@
...
+  // Grab the News Feed Posts that the user has retrieved from the store
   getNewsFeedPosts() {
     this.setState({
-      posts: NewsFeedStore.posts
+      contentFeed: NewsFeedStore.contentFeed,
     });
   }
 
+  getTextArea() {
+    return document.getElementsByClassName("user-post-form")[0].getElementsByTagName('textarea')[0].value;
```
Comment (larajanecka) [inline src/js/pages/attendee/NewsFeed.jsx]: We should probably have the text area be attacked the the component state and just have an onChange function that updates the state.

File: src/js/pages/attendee/NewsFeed.jsx [hunk 5]
```diff
@@ -1,130 +1,210 @@
...
-              <div style={alignRight}>
-                <button class="submit-form btn btn-primary" style={aright} onClick={this.createPost}>Make a post</button>
+      <div>
+        <h1>News Feed</h1>
+        <div className="newsFeed">
+          <div className="user-post">
+            <form className="user-post-form">
+              <div className="input">
+                <div className="left-input inline">
+                  <textarea className="text-input border"
```
Comment (larajanecka) [inline src/js/pages/attendee/NewsFeed.jsx]: Have this populated with this.state.new_post_content (or some such name) and add an onchange listener that grabs the value of the event target and updates the state. This way we dont need that super convoluted element look up in getTextArea and just have the value we need.

File: src/js/stores/NewsFeedStore.jsx [hunk 6]
```diff
@@ -78,22 +84,21 @@ class NewsFeedStore extends EventEmitter {
...
-        // TODO: Readd this back in
-        this.emit("updated");
+        // TODO: ensure that client sees their post on the front end
+        this.emit("post");
         break;
       }
       case "NEWSFEED_GET": {
-        this.posts = action.posts;
-        // TODO: Append posts, don't simply delete old data
+        Array.prototype.push.apply(this.contentFeed.items, action.response.data.posts[0]);
```
Comment (larajanecka) [inline src/js/stores/NewsFeedStore.jsx]: I don't know how the back end works, but this seems potentially sketchy. Will the get ever return more than one post?

File: src/js/components/NewsFeedPost.jsx [hunk 7]
```diff
@@ -25,98 +22,58 @@ export default class NewsFeedPost extends React.Component {
...
-          <span class="text" style={font30}>{text} </span>
+      <div className="newsfeed-post">
+        <div className="user-photo-container inline">
+          <img src={poster_icon} className="user-photo"/>
+        </div>
+        <div className="post-text inline">
+          <p className="username">{poster_name}</p>
+          <p className="text">{text}</p>
+          <div className="social-bar">
+            <FaHeart className="heart-icon clickable" onClick={this.changeLike} style={heartColor} size={20}/>
```
Comment (larajanecka) [inline src/js/components/NewsFeedPost.jsx]: @schneider-clarisse should we just have a class name that we toggle, or is inline more correct? Not sure what the react norms for this are.

File: src/js/pages/attendee/NewsFeed.jsx [hunk 8]
```diff
@@ -1,130 +1,210 @@
...
-      width: "49%",
-      display: "inline-block",
-      marginTop: "-30px",
-    }
-
-    var aright = {
-      float: "right",
+    let paginate = (<div className="default-cursor">End of Content</div>);
+    if (contentFeed.nextOffset) {
+      paginate = (<div className="clickable link" onClick={NewsFeedStore.get}>Load more...</div>);
```
Comment (larajanecka) [inline src/js/pages/attendee/NewsFeed.jsx]: Pagination, nice.

**Review by larajanecka — CHANGES_REQUESTED** [review]

File: src/js/pages/attendee/NewsFeed.jsx (hunk 2 above)
Comment (ericljpemberton) [reply src/js/pages/attendee/NewsFeed.jsx:12]: Is it already that style in other places? I thought I was following convention. Otherwise we can do 1 app wide change for all instances of this

File: src/js/pages/attendee/NewsFeed.jsx (hunk 3 above)
Comment (ericljpemberton) [reply src/js/pages/attendee/NewsFeed.jsx]: It might be in auth0 or we might consider putting it there?

File: src/js/stores/NewsFeedStore.jsx (hunk 6 above)
Comment (ericljpemberton) [reply src/js/stores/NewsFeedStore.jsx]: I'm not a huge fan of how I did this either and am very open to other ways of paginating. I haven't really implemented it before. The GET will almost always get more than 1 post but push.apply pretty much appends one array to the other. (and unlike concat, it doesn't create a new array)
https://davidwalsh.name/merge-arrays-javascript

File: src/js/components/NewsFeedPost.jsx (hunk 7 above)
Comment (ericljpemberton) [reply src/js/components/NewsFeedPost.jsx]: Not positive what my thought process for this was, I may have been following
https://github.com/Calligre/attendee-web/blob/32e685e019a1bf1bef56df69b85c63a8fb23b96f/src/js/components/Events.jsx#L22
?

**Comment by ericljpemberton** [comment]
@larajanecka @StevenLocke If you get a chance to look this over I'd appreciate it. Posting photos is not quite working yet. Still getting url issues but everything else should be functional. Only 1 or two lines of code should change when everything is sorted out anyway

I ran the linter on NewsFeed and NewsFeed post. There are still errors but the basic stuff has been corrected. Mostly things like declaring stores and components in npm, props validation and img with declared roles are still incomplete. I'll try to look into that tomorrow

File: src/js/pages/attendee/NewsFeed.jsx (hunk 2 above)
Comment (larajanecka) [reply src/js/pages/attendee/NewsFeed.jsx:12]: Some places use it and some dont. Once we get mvp on organizer app I want to do a bunch of these app wide changes. No need to change it now, just wanted to remind myself to do it in the future.

File: src/js/pages/attendee/NewsFeed.jsx (hunk 3 above)
Comment (larajanecka) [reply src/js/pages/attendee/NewsFeed.jsx]: @kyl191 might know more. There was similar discussion around the preferences (since we allow organizers to turn off integrations in the conference settings).

File: src/js/stores/NewsFeedStore.jsx (hunk 6 above)
Comment (larajanecka) [reply src/js/stores/NewsFeedStore.jsx]: Ok. Sounds like this will work then. Neat.

File: src/js/components/NewsFeedPost.jsx [hunk 9]
```diff
@@ -1,59 +1,67 @@
...
+  // componentWillUnmount() {}
 
   changeLike() {
     if (this.state.current_user_likes) {
-      NewsFeedStore.decrementLike();
+      NewsFeedStore.unlikePost(this.state.id);
+      // TODO: Potentially move this logic to be based on emitted action?
       this.setState({
         current_user_likes: false,
+        like_count: (parseInt(this.state.like_count) - 1).toString(),
```
Comment (larajanecka) [inline src/js/components/NewsFeedPost.jsx]: Why the - 1?

File: src/js/components/NewsFeedPost.jsx [hunk 10]
```diff
@@ -1,59 +1,67 @@
...
+      NewsFeedStore.likePost(this.state.id);
       this.setState({
         current_user_likes: true,
+        like_count: (parseInt(this.state.like_count) + 1).toString(),
       });
     }
+  }
 
+  retweet() {
+    const retweetText = '"' + this.state.text + '" - ' + this.state.poster_name + '\n';
```
Comment (larajanecka) [inline src/js/components/NewsFeedPost.jsx]: Could use string interpolation to make this a bit easier (wrap in `).

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

File: src/js/stores/NewsFeedStore.jsx [hunk 11]
```diff
@@ -3,36 +3,39 @@ import AuthService from "util/AuthService";
...
-      //   "limit": "2"
-      //   "offset": offset
-      // },
+      type: "GET",
+      url: url + "/api/social",
+      data: params,
+      dataType: "json",
       cache: false,
       success: function(response) {
+        console.log(response);
```
Comment (larajanecka) [inline src/js/stores/NewsFeedStore.jsx]: Don't think we need this log anymore.

File: src/js/stores/NewsFeedStore.jsx [hunk 12]
```diff
@@ -52,25 +55,58 @@ class NewsFeedStore extends EventEmitter {
...
+  likePost(postId) {
+    $.ajax({
+      type: "POST",
+      url: url + "/api/social/" + postId + "/likes",
+      headers: {
+        "Authorization": "Bearer " + AuthService.getToken()
+      },
+      contentType:"application/json",
+      cache: false,
+      statusCode: {
```
Comment (larajanecka) [inline src/js/stores/NewsFeedStore.jsx]: ooohhhhh I didn't know this was a thing. We need to use status code more.

File: src/js/stores/NewsFeedStore.jsx [hunk 13]
```diff
@@ -52,25 +55,58 @@ class NewsFeedStore extends EventEmitter {
...
+        500: function() {
+          alert("Unable to perform this action, please try again later");
+        }
+      },
+      success: function(response) {
+        console.log("LIKE SUCCESS")
```
Comment (larajanecka) [inline src/js/stores/NewsFeedStore.jsx]: kill excess logs to prevent cluttering.

File: src/js/stores/NewsFeedStore.jsx [hunk 14]
```diff
@@ -52,25 +55,58 @@ class NewsFeedStore extends EventEmitter {
...
+        dispatcher.dispatch({type: "POST_LIKE", response: response});
+      },
+      failure: function(error) {
+        console.log("LIKE FAILURE")
```
Comment (larajanecka) [inline src/js/stores/NewsFeedStore.jsx]: kill

File: src/js/stores/NewsFeedStore.jsx [hunk 15]
```diff
@@ -79,19 +115,79 @@ class NewsFeedStore extends EventEmitter {
       contentType:"application/json",
       cache: false,
       success: function(response) {
-        dispatcher.dispatch({type: "NEWSFEED_POST", post: response["id"]});
+        console.log("THIS ACTUALLY FINISHED");
```
Comment (larajanecka) [inline src/js/stores/NewsFeedStore.jsx]: kill

**Review by larajanecka — APPROVED** [review]
Couple of small nit picks, but nothing that should block it.

File: src/js/stores/NewsFeedStore.jsx (hunk 15 above)
Comment (ericljpemberton) [reply src/js/stores/NewsFeedStore.jsx]: Currently necessary for me to test the POSTing after Kyle looks at it

Comment (ericljpemberton) [reply src/js/stores/NewsFeedStore.jsx]: (That applies to almost all of those console statements. They'll be cleaned up as soon as the photo posting is functional)

File: src/js/components/NewsFeedPost.jsx (hunk 9 above)
Comment (ericljpemberton) [reply src/js/components/NewsFeedPost.jsx]: This is the process of the user unliking the post. So I remove one from the count on the post if they had previously liked it. I was wondering if I should do this optimistically as above so the user sees feedback or actually dispatch it when I get a response and make the change then

File: src/js/stores/NewsFeedStore.jsx (hunk 15 above)
Comment (larajanecka) [reply src/js/stores/NewsFeedStore.jsx]: kk.

File: src/js/pages/attendee/NewsFeed.jsx (hunk 3 above)
Comment (kyl191) [reply src/js/pages/attendee/NewsFeed.jsx]: I talked with Kevin, then took some time and poked around this.
Turns out as much as I would like it to be, Auth0 is not magic fairy dust that solves all our auth and user info problems.

tl;dr is we're going to have an API to hit - probably going to extend `/me`.

File: src/js/stores/NewsFeedStore.jsx (hunk 6 above)
Comment (kyl191) [reply src/js/stores/NewsFeedStore.jsx]: My only concern here is that it's `posts_[0]_` - posts is an array, why are we only taking the first post?
